### PR TITLE
O3 2168 Form Engine: Handle case when # of concepts on a form exceeds REST page length

### PR DESCRIPTION
**Pull request by sibendire**
## Requirements

- [ ] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-dev.docs.openmrs.org/#/getting_started/contributing?id=your-pr-title-should-indicate-the-type-of-change-it-is) label. See existing PR titles for inspiration.
- [ ] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://zeroheight.com/23a080e38/p/880723-introduction).
- [ ] My work includes tests or is validated by existing tests.
## Description of what I changed
Enhanced the concept fetching process for form references to ensure accurate retrieval of all concepts. Addressed issues related to REST API pagination that were causing missed concepts. Implemented a more precise fetching logic that aligns with the API's pagination mechanism.
## Related Issue
https://issues.openmrs.org/browse/O3-2168

### CODE REVIEW

**Comment by sibendire** [comment]
Hello, @dkayiwa could you guide me on how to fix this issue and run the tests locally?
Thank you.

File: src/hooks/useConcepts.tsx [hunk 1]
```diff
@@ -5,10 +5,31 @@ const conceptRepresentation =
...
+
+    for (let page = 2; page <= totalPages; page++) {
+      const nextPageUrl = `${url}&startIndex=${(page - 1) * pageSize}&v=${conceptRepresentation}`;
+      const nextPageResponse = await openmrsFetch(nextPageUrl);
+      const nextPageResults: OpenmrsResource[] = await nextPageResponse.json();
+      fetchedResults = fetchedResults.concat(nextPageResults);
+    }
+
+    return { results: fetchedResults };
+  };
```
Comment (samuelmale) [inline src/hooks/useConcepts.tsx:26]: This function seems to be redundant like it's not being referenced anywhere.

File: src/hooks/useConcepts.tsx [hunk 2]
```diff
@@ -5,10 +5,31 @@ const conceptRepresentation =
...
-  // TODO: handle paging (ie when number of concepts greater than default limit per page)
-  const { data, error, isLoading } = useSWRImmutable<{ data: { results: Array<OpenmrsResource> } }, Error>(
+  const pageSize = 50;
+   // TODO: handle paging (ie when number of concepts greater than default limit per page)
+   const { data, error, isLoading } = useSWRImmutable<{ data: { results: Array<OpenmrsResource> } }, Error>(
     `/ws/rest/v1/concept?references=${Array.from(references).join(',')}&v=${conceptRepresentation}`,
     openmrsFetch,
   );
+
+  const fetchConcepts = async (url:string) => {
```
Comment (mogoodrich) [inline src/hooks/useConcepts.tsx:15]: Yes, agree with @samuelmale ... this function never seems to be called, @sibendire  have you tested that this actually works?

Comment (mogoodrich) [reply src/hooks/useConcepts.tsx:15]: It also looks like we are no longer using SWR... we would want to preserve the usage of SWR here....

**Review by mogoodrich — CHANGES_REQUESTED** [review]

File: src/hooks/useConcepts.tsx (hunk 2 above)
Comment (mogoodrich) [reply src/hooks/useConcepts.tsx:15]: No worries @sibendire  but I don't see any new changes?

Comment (sibendire) [reply src/hooks/useConcepts.tsx]: Hello @mogoodrich  I believe with your guidance I can make as it expect. Could you review just slight change made and give me the feedback on how I have been missing up to fix the issue ?

Thank you.

Comment (sibendire) [reply src/hooks/useConcepts.tsx]: Am looking forward to @mogoodrich @samuelmale and @denniskigen for your response and guidance that will teach me more about this.
Thank you.

**Pull request closed without merging**